src/twister2/environment/environment.py

```diff
@@ -0,0 +1,119 @@
+    """
+    When TwisterConfig is generated first time, then information about used
+    toolchain version should be taken by CMake script and then saved into
+    environment_info.json file. In case when xdist is used, then every time when
+    TwisterConfig is generated once again by each worker, information about used
+    toolchain version can be taken from environment_info.json file to avoid
+    calling CMake script several times in row.
+    """
+    environment_info_file_name = 'environment_info.json'
+    environment_info_file_path: Path = Path(output_dir) / environment_info_file_name
+
+    if environment_info_file_path.is_file():
+        used_toolchain_version = _get_toolchain_version_from_env_info_file(environment_info_file_path)
+    else:
+        used_toolchain_version = _get_toolchain_version_from_cmake_script(zephyr_base)
+        _save_toolchain_version_to_env_info_file(environment_info_file_path, used_toolchain_version)
+
+    return used_toolchain_version
+
+
+def _get_toolchain_version_from_env_info_file(file_path: Path) -> str:
+    try:
+        with open(file_path, 'r') as file:
+            environment_info = json.load(file)
+            used_toolchain_version = environment_info['used_toolchain_version']
+    except Exception:
+        logger.error('Problem with get info about used toolchain version.')
+        raise
+    return used_toolchain_version
+
+
+def _save_toolchain_version_to_env_info_file(file_path: Path, toolchain_version: str) -> None:
+    environment_info = {'used_toolchain_version': toolchain_version}
+    with open(file_path, 'w') as file:
+        json.dump(environment_info, file, indent=4)
+
+
+def _get_toolchain_version_from_cmake_script(zephyr_base: str) -> str:
+    """
+    TODO: refactor and simplify this function
+    TODO: write unit tests dedicated for this function
+    """
+    toolchain_script = Path(zephyr_base) / Path('cmake/verify-toolchain.cmake')
+    result = _run_cmake_script([toolchain_script, 'FORMAT=json'])
+
+    try:
+        if result['returncode']:
+            raise TwisterFatalError(f"E: {result['returnmsg']}")
+    except Exception as e:
+        print(str(e))
+        sys.exit(2)
+    toolchain_version = json.loads(result['stdout'])['ZEPHYR_TOOLCHAIN_VARIANT']
+    logger.info(f"Using '{toolchain_version}' toolchain.")
+    return toolchain_version
+
+
+def _run_cmake_script(args=[]):
+    """
+    TODO: refactor and simplify this function
+    TODO: write unit tests dedicated for this function
+    """
+    script = os.fspath(args[0])
+
+    logger.debug('Running cmake script %s', script)
+
+    cmake_args = ['-D{}'.format(a.replace('"', '')) for a in args[1:]]
+    cmake_args.extend(['-P', script])
+
+    cmake = shutil.which('cmake')
+    if not cmake:
+        msg = 'Unable to find `cmake` in path'
+        logger.error(msg)
+        raise Exception(msg)
+    cmd = [cmake] + cmake_args
+    log_command(logger, 'Calling cmake', cmd)
+
+    kwargs = dict()
```

Comment:
I don't see any reason to keep Popen arguments in dictionary, they can be pass to the class directly.

Reply:
Changed